src/crypto/hashing/merkle.rs

```diff
@@ -81,20 +81,93 @@ impl MerkleTree {
             }
             // If there's one node left over, add it to the end
             if prev_layer.len() % 2 == 1 {
-                current_layer.push(prev_layer.len() - 1);
+                current_layer.push(prev_layer[prev_layer.len() - 1]);
             }
         }
         MerkleTree {
             nodes,
             root: current_layer[0],
-            leaves: leaves.len(),
+            size: leaves.len(),
         }
     }
+
+    /// Generate a proof that the given item is contained within the Merkle tree.
+    pub fn construct_proof(&self, index: usize) -> Vec<Hash> {
+        if index >= self.size {
+            panic!()
+        } else {
+            let mut proof = vec![];
+            let mut relative_index = index;
+            let mut current = &self.nodes[self.root];
+            while current.size > 1 {
+                let children = current
+                    .children
+                    .map(|(left, right)| (&self.nodes[left], &self.nodes[right]))
+                    .unwrap();
+                if relative_index < children.0.size {
+                    proof.push(children.1.value);
+                    current = children.0;
+                } else {
+                    relative_index -= children.0.size;
+                    proof.push(children.0.value);
+                    current = children.1;
+                }
+            }
+            proof
+        }
+    }
+
+    /// Verify a proof that a given Merkle tree contains the given leaf node at the given index
+    pub fn verify_proof<T: Hashable>(
+        index: usize,
+        size: usize,
+        leaf: T,
+        tree_hash: Hash,
+        proof: &[Hash],
+    ) -> bool {
+        verify_proof_rec(index, size, leaf.hash(), proof)
+            .map(|found| hash![found, size] == tree_hash)
+            .unwrap_or(false)
+    }
+}
+
+/// Underlying function to recursively verify a proof
+fn verify_proof_rec(index: usize, size: usize, leaf: Hash, proof: &[Hash]) -> Option<Hash> {
+    let left_size = left_child_size(size);
+    if size == 1 {
+        Some(leaf)
+    } else if proof.is_empty() {
+        None
+    } else if index < left_size {
+        verify_proof_rec(index, left_size, leaf, &proof[1..]).map(|left| hash![left, proof[0]])
+    } else {
+        let relative_index = index - left_size;
+        verify_proof_rec(relative_index, size - left_size, leaf, &proof[1..])
+            .map(|right| hash![proof[0], right])
+    }
+}
+
+/// Gets the size of the left child, given the size of a parent MerkleNode
+fn left_child_size(mut size: usize) -> usize {
```

Comment:
left_child_size can be further optimized from logarithmic time to constant time by bit manipulation